### PR TITLE
Istio: Update to 1.6.14

### PR DESCRIPTION
**Pull request by jrajahalme**
Update Istio integration to Istio release 1.6.14 now that 1.5.x is no longer supported.

Istio CI test is now unquarantined on K8s 1.19 and K8s 1.20.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

```release-note
Istio integration is updated to Istio release 1.6.14
```

### CODE REVIEW

**Comment by jrajahalme** [comment]
test-me-please

**Comment by jrajahalme** [comment]
Using REDIRECT instead of TPROXY has a scenario where the original source address is lost. Specifically, if two connections from the same source are received for the same destination port, REDIRECT will do a source NAT to ensure unique 5-tuples. This will lead to incorrect source security ID mapping.

Upstream Istio is fixing TPROXY (https://github.com/istio/istio/pull/28457), will keep this as draft for now.

**Comment by jrajahalme** [comment]
test-me-please

**Comment by jrajahalme** [comment]
test-me-please

**Comment by jrajahalme** [comment]
retest-gke

**Comment by jrajahalme** [comment]
test-me-please

**Comment by jrajahalme** [comment]
retest-4.9

**Comment by jrajahalme** [comment]
retest-4.9

**Comment by jrajahalme** [comment]
test-me-please

**Comment by jrajahalme** [comment]
retest-4.9

**Comment by jrajahalme** [comment]
retest-netnext

**Comment by jrajahalme** [comment]
retest-net-next

**Comment by jrajahalme** [comment]
test-me-please

**Comment by jrajahalme** [comment]
test-only --focus="K8sIstio"

**Comment by jrajahalme** [comment]
test-only --focus="K8sIstio"

**Comment by jrajahalme** [comment]
test-only --focus="K8sIstio"

**Comment by jrajahalme** [comment]
test-me-please

**Comment by jrajahalme** [comment]
test-only --focus="K8sIstio"

**Comment by jrajahalme** [comment]
test-me-please

**Comment by jrajahalme** [comment]
test-only --focus="K8sIstio"

**Comment by jrajahalme** [comment]
test-me-please

**Comment by jrajahalme** [comment]
retest-gke

**Comment by jrajahalme** [comment]
test-only --focus="K8sIstio"

**Comment by jrajahalme** [comment]
test-me-please

**Comment by jrajahalme** [comment]
retest-gke

File: Documentation/gettingstarted/istio.rst [hunk 1]
```diff
@@ -29,26 +37,26 @@ Step 2: Install cilium-istioctl
...
-     curl -L https://github.com/cilium/istio/releases/download/1.5.9/cilium-istioctl-1.5.9-osx.tar.gz | tar xz
+     curl -L https://github.com/cilium/istio/releases/download/1.6.14/cilium-istioctl-1.6.14-osx.tar.gz | tar xz
 
 .. note::
 
    Cilium integration, as presented in this Getting Started Guide, has
-   been tested with Kubernetes releases 1.14, 1.15, 1.16, 1.17, and
-   1.18. Note that this does *not* work with K8s 1.13.
+   been tested with Kubernetes releases 1.14, 1.15, 1.16, 1.17, 1.18,
+   1.19, and 1.20. Note that this does *not* work with K8s 1.13.
```
Comment (aditighag) [inline Documentation/gettingstarted/istio.rst:59]: I'm just curious -- what's so (un)special about 1.13?

File: test/k8sT/istio.go [hunk 2]
```diff
@@ -29,46 +29,51 @@ import (
...
+			// "kubeProxyReplacement": "disabled",
+			// "debug.enabled": "true",
+			// "debug.verbose": "flow",
 		}
 
-		// Map of tested runtimes for cilium-istioctl
+		// Map for tested cilium-istioctl release targets if not GOOS-GORACH
 		ciliumIstioctlOSes = map[string]string{
-			"darwin": "osx",
-			"linux":  "linux",
```
Comment (aditighag) [inline test/k8sT/istio.go:60]: Is this change intentional?

**Review by aditighag — COMMENTED** [review]
Perhaps it'll be helpful to add more details to the commit description around why the test was quarantined before, and how it was fixed with 1.16?

File: test/k8sT/istio.go [hunk 3]
```diff
@@ -29,46 +29,51 @@ import (
...
-		// " --set values.proxy_init.image=docker.io/cilium/istio_proxy:1.5.9"
+		// " --set values.pilot.image=docker.io/cilium/istio_pilot:1.6.14" +
+		// " --set values.global.proxy.image=docker.io/cilium/istio_proxy:1.6.14" +
+		// " --set values.global.proxy_init.image=docker.io/cilium/istio_proxy:1.6.14" +
+		// " --set values.global.proxy.logLevel=trace"
+		// " --set values.global.logging.level=debug"
+		// " --set values.global.mtls.auto=false"
 		ciliumOptions = map[string]string{
 			// "proxy.sidecarImageRegex": "jrajahalme/istio_proxy",
+			// "kubeProxyReplacement": "disabled",
```
Comment (aditighag) [inline test/k8sT/istio.go:57]: This should be enabled?

**Review by aditighag — APPROVED** [review]
I was able to successfully run the test locally.

File: Documentation/gettingstarted/istio.rst [hunk 4]
```diff
@@ -8,19 +8,27 @@
...
+deployed without Cilium integration by running a standard version of
+``istioctl``.  In that case Cilium will enforce HTTP L7 policies outside
+of the Istio sidecar proxy.
 
 .. include:: gsg_requirements.rst
 
 .. note::
 
-   If running on minikube, you may need to up the memory and CPUs
+   If running on minikube, make sure it is using a VM, as the ingress
```
Comment (kkourt) [inline Documentation/gettingstarted/istio.rst]: Not sure what is meant by ingress here, ingress rules? ingress traffic?
It might be obvious to people with this use case, but might be worth specifying.

File: test/k8sT/istio.go [hunk 5]
```diff
@@ -316,12 +362,53 @@ var _ = SkipContextIf(func() bool {
...
+				return allGood
+			}, "Istio sidecar proxies are not configured", &helpers.TimeoutConfig{Timeout: helpers.HelperTimeout})
+			Expect(err).Should(BeNil(), "Cannot configure Istio sidecar proxies")
+
+			By("Testing service local access")
+
+			err = helpers.WithTimeout(func() bool {
+				allGood := true
+
+				allGood = shouldConnect(reviewsPodV1.String(), "reviews", formatLocalAPI(apiPort, health)) && allGood
```
Comment (kkourt) [inline test/k8sT/istio.go]: AFAICT, `shouldConnect` does the same as `shouldCurlConnect` using wget because some containers have wget but others have curl. When I originally read the code I thought one test was a HTTP connect while the other one was a TCP connect (e.g., using nc). Maybe it's worth renaming `shouldConnect` to `shouldWgetConnect`?

File: test/k8sT/istio.go [hunk 6]
```diff
@@ -316,12 +362,53 @@ var _ = SkipContextIf(func() bool {
...
+				allGood = shouldHaveService(ratingsPodV1.String(), ratings, apiPort, inbound) && allGood
+				allGood = shouldHaveService(productpagePodV1.String(), details, apiPort, outbound) && allGood
+				allGood = shouldHaveService(detailsPodV1.String(), details, apiPort, inbound) && allGood
+				allGood = shouldHaveService(productpagePodV1.String(), ratings, apiPort, outbound) && allGood
+
+				return allGood
+			}, "Istio sidecar proxies are not configured", &helpers.TimeoutConfig{Timeout: helpers.HelperTimeout})
+			Expect(err).Should(BeNil(), "Cannot configure Istio sidecar proxies")
+
+			By("Testing service local access")
```
Comment (kkourt) [inline test/k8sT/istio.go:387]: So this test will execute the commands inside the pod that each service runs to ensure that everything is OK? (there is no Istio involved here at all.)

**Review by aanm — APPROVED** [review]
Nothing else from my side besides what Kornilios and Aditi have reviewed. @jrajahalme do we need to backport any changes for 1.9?

File: test/k8sT/istio.go [hunk 7]
```diff
@@ -316,12 +362,53 @@ var _ = SkipContextIf(func() bool {
...
+				allGood := true
+
+				allGood = shouldHaveService(reviewsPodV1.String(), ratings, apiPort, outbound) && allGood
+				allGood = shouldHaveService(ratingsPodV1.String(), ratings, apiPort, inbound) && allGood
+				allGood = shouldHaveService(productpagePodV1.String(), details, apiPort, outbound) && allGood
+				allGood = shouldHaveService(detailsPodV1.String(), details, apiPort, inbound) && allGood
+				allGood = shouldHaveService(productpagePodV1.String(), ratings, apiPort, outbound) && allGood
+
+				return allGood
+			}, "Istio sidecar proxies are not configured", &helpers.TimeoutConfig{Timeout: helpers.HelperTimeout})
```
Comment (christarazi) [inline test/k8sT/istio.go]: This should be an `Eventually`. We had a PR by Paul recently that converted `helpers.WithTimeout` and other variants to the native Ginkgo versions.

File: test/k8sT/istio.go [hunk 8]
```diff
@@ -316,12 +362,53 @@ var _ = SkipContextIf(func() bool {
...
+			Expect(err).Should(BeNil(), "Cannot configure Istio sidecar proxies")
+
+			By("Testing service local access")
+
+			err = helpers.WithTimeout(func() bool {
```
Comment (christarazi) [inline test/k8sT/istio.go]: Same here

**Review by christarazi — CHANGES_REQUESTED** [review]
Adding onto others' comments, this one more focused on CI.

File: test/k8sT/istio.go [hunk 9]
```diff
@@ -29,46 +29,51 @@ import (
...
+		// " --set values.global.mtls.auto=false"
 		ciliumOptions = map[string]string{
 			// "proxy.sidecarImageRegex": "jrajahalme/istio_proxy",
+			// "kubeProxyReplacement": "disabled",
+			// "debug.enabled": "true",
+			// "debug.verbose": "flow",
 		}
 
-		// Map of tested runtimes for cilium-istioctl
+		// Map for tested cilium-istioctl release targets if not GOOS-GORACH
```
Comment (qmonnet) [inline test/k8sT/istio.go]:
```suggestion
		// Map for tested cilium-istioctl release targets if not GOOS-GOARCH
```

**Review by qmonnet — APPROVED** [review]
Not much to add to the other reviews. If anything, don't hesitate to split such PRs into several commits in the future, it might help with review and history :).

File: test/k8sT/istio.go (hunk 2 above)
Comment (jrajahalme) [reply test/k8sT/istio.go:60]: Yes, due to arm64 support in Istioctl releases. Will likely change soon in future releases again when macOS gets arm64 builds as well.

File: test/k8sT/istio.go (hunk 3 above)
Comment (jrajahalme) [reply test/k8sT/istio.go:57]: It is left to the whatever the default is for the current integration (e.g., GKE). I left it commented there as a reminder for testing future releases as there may be significant interaction between this setting and policy enforcement with Istio sidecars.

File: test/k8sT/istio.go (hunk 6 above)
Comment (jrajahalme) [reply test/k8sT/istio.go:387]: Right. Sidecar redirection in the TPROXY mode is broken in the upstream Istio 1.6.14 release. I added these tests to rule out issues with reaching the local service from the sidecar container.

File: Documentation/gettingstarted/istio.rst (hunk 4 above)
Comment (jrajahalme) [reply Documentation/gettingstarted/istio.rst]: Changed this to `If running on minikube, make sure it is using a VM, as Istio ingress gateway may not be reachable from your host otherwise.`

File: test/k8sT/istio.go (hunk 5 above)
Comment (jrajahalme) [reply test/k8sT/istio.go]: Done. But left `shouldNotConnect()` as-is.

File: test/k8sT/istio.go (hunk 7 above)
Comment (jrajahalme) [reply test/k8sT/istio.go]: Done.

File: test/k8sT/istio.go (hunk 8 above)
Comment (jrajahalme) [reply test/k8sT/istio.go]: Ditto.

File: test/k8sT/istio.go (hunk 9 above)
Comment (jrajahalme) [reply test/k8sT/istio.go]: Fixed, thanks!

**Comment by jrajahalme** [comment]
test-me-please

**Review by christarazi — APPROVED** [review]
